File: jsimple-oauth/src/main/java/jsimple/oauth/builder/ServiceBuilder.java
/*
 * Portions copyright (c) 2012-2014 Microsoft Mobile.  All Rights Reserved.
 *
 * This file is based on or incorporates material from the Scribe Java OAuth
 * library https://github.com/fernandezpablo85/scribe-java (collectively, "Third
 * Party Code"). Microsoft Mobile is not the original author of the Third Party
 * Code.
 *
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package jsimple.oauth.builder;

import jsimple.oauth.builder.api.OAuthApi;
import jsimple.oauth.model.OAuthConfig;
import jsimple.oauth.model.OAuthConstants;
import jsimple.oauth.model.OAuthLogger;
import jsimple.oauth.model.SignatureType;
import jsimple.oauth.oauth.OAuthService;
import org.jetbrains.annotations.Nullable;


/**
 * Implementation of the Builder pattern, with a fluent interface that creates a {@link OAuthService}
 *
 * @author Pablo Fernandez
 */
public class ServiceBuilder {
    private @Nullable String serviceApiKey;
    private @Nullable String serviceApiSecret;
    private String serviceCallback;
    private @Nullable OAuthApi api;
    private @Nullable String serviceScope;
    private SignatureType serviceSignatureType;
    private @Nullable OAuthLogger serviceDebugLogger;

    /**
     * Default constructor
     */
    public ServiceBuilder() {
        this.serviceCallback = OAuthConstants.OUT_OF_BAND;
        this.serviceSignatureType = SignatureType.Header;
        this.serviceDebugLogger = null;
    }

    /**
     * Configures the {@link jsimple.oauth.builder.api.OAuthApi}
     * <p/>
     * Overloaded version. Let's you use an instance instead of a class.
     *
     * @param api instance of {@link jsimple.oauth.builder.api.OAuthApi}s
     * @return the {@link ServiceBuilder} instance for method chaining
     */
    public ServiceBuilder provider(OAuthApi api) {
        assert api != null : "Api cannot be null";
        this.api = api;
        return this;
    }

    /**
     * Adds an OAuth callback url
     *
     * @param callback callback url. Must be a valid url or 'oob' for out of band OAuth
     * @return the {@link ServiceBuilder} instance for method chaining
     */
    public ServiceBuilder callback(String callback) {
        assert callback != null : "Callback can't be null";
        this.serviceCallback = callback;
        return this;
    }

    /**
     * Configures the api key
     *
     * @param apiKey The api key for your application
     * @return the {@link ServiceBuilder} instance for method chaining
     */
    public ServiceBuilder apiKey(String apiKey) {
        assert !apiKey.isEmpty() : "Invalid Api key";
        this.serviceApiKey = apiKey;
        return this;
    }

    /**
     * Configures the api secret
     *
     * @param apiSecret The api secret for your application
     * @return the {@link ServiceBuilder} instance for method chaining
     */
    public ServiceBuilder apiSecret(String apiSecret) {
        assert !apiSecret.isEmpty() : "Invalid Api secret";
        this.serviceApiSecret = apiSecret;
        return this;
    }

    /**
     * Configures the OAuth scope. This is only necessary in some APIs (like Google's).
     *
     * @param scope The OAuth scope
     * @return the {@link ServiceBuilder} instance for method chaining
     */
    public ServiceBuilder scope(String scope) {
        assert !scope.isEmpty() : "Invalid OAuth scope";
        this.serviceScope = scope;
        return this;
    }

    /**
     * Configures the signature type, choose between header, querystring, etc. Defaults to Header
     *
     * @param type signature type
     * @return the {@link ServiceBuilder} instance for method chaining
     */
    public ServiceBuilder signatureType(SignatureType type) {
        this.serviceSignatureType = type;
        return this;
    }

    public ServiceBuilder debugLogger(OAuthLogger logger) {
        assert logger != null : "debug stream can't be null";
        this.serviceDebugLogger = logger;
        return this;
    }

    /*
    public ServiceBuilder debug() {
        this.debugLogger(System.out);
        return this;
    }
    */

    /**
     * Returns the fully configured {@link OAuthService}
     *
     * @return fully configured {@link OAuthService}
     */
    public OAuthService build() {
        assert api != null : "nullness";          // You must specify a valid api through the provider() method"
        assert serviceApiKey != null : "nullness";       // You must provide an api key
        assert serviceApiSecret != null : "nullness";    // You must provide an api secret
        return api.createService(new OAuthConfig(serviceApiKey, serviceApiSecret, serviceCallback, serviceSignatureType, serviceScope, serviceDebugLogger));
    }
}

File: jsimple-oauth/c#/jsimple/oauth/builder/ServiceBuilder.cs
using System.Diagnostics;

namespace jsimple.oauth.builder {

    using OAuthApi = jsimple.oauth.builder.api.OAuthApi;
    using OAuthConfig = jsimple.oauth.model.OAuthConfig;
    using OAuthConstants = jsimple.oauth.model.OAuthConstants;
    using OAuthLogger = jsimple.oauth.model.OAuthLogger;
    using SignatureType = jsimple.oauth.model.SignatureType;
    using OAuthService = jsimple.oauth.oauth.OAuthService;



    /// <summary>
    /// Implementation of the Builder pattern, with a fluent interface that creates a <seealso cref="OAuthService"/>
    /// 
    /// @author Pablo Fernandez
    /// </summary>
    public class ServiceBuilder {
        private string apiKey_Renamed;
        private string apiSecret_Renamed;
        private string callback_Renamed;
        private OAuthApi api;
        private string scope_Renamed;
        private SignatureType signatureType_Renamed;
        private OAuthLogger debugLogger_Renamed;

        /// <summary>
        /// Default constructor
        /// </summary>
        public ServiceBuilder() {
            this.callback_Renamed = OAuthConstants.OUT_OF_BAND;
            this.signatureType_Renamed = SignatureType.Header;
            this.debugLogger_Renamed = null;
        }

        /// <summary>
        /// Configures the <seealso cref="jsimple.oauth.builder.api.OAuthApi"/>
        /// <p/>
        /// Overloaded version. Let's you use an instance instead of a class.
        /// </summary>
        /// <param name="api"> instance of <seealso cref="jsimple.oauth.builder.api.OAuthApi"/>s </param>
        /// <returns> the <seealso cref="ServiceBuilder"/> instance for method chaining </returns>
        public virtual ServiceBuilder provider(OAuthApi api) {
            Debug.Assert(api != null, "Api cannot be null");
            this.api = api;
            return this;
        }

        /// <summary>
        /// Adds an OAuth callback url
        /// </summary>
        /// <param name="callback"> callback url. Must be a valid url or 'oob' for out of band OAuth </param>
        /// <returns> the <seealso cref="ServiceBuilder"/> instance for method chaining </returns>
        public virtual ServiceBuilder callback(string callback) {
            Debug.Assert(callback != null, "Callback can't be null");
            this.callback_Renamed = callback;
            return this;
        }

        /// <summary>
        /// Configures the api key
        /// </summary>
        /// <param name="apiKey"> The api key for your application </param>
        /// <returns> the <seealso cref="ServiceBuilder"/> instance for method chaining </returns>
        public virtual ServiceBuilder apiKey(string apiKey) {
            Debug.Assert(apiKey.Length > 0, "Invalid Api key");
            this.apiKey_Renamed = apiKey;
            return this;
        }

        /// <summary>
        /// Configures the api secret
        /// </summary>
        /// <param name="apiSecret"> The api secret for your application </param>
        /// <returns> the <seealso cref="ServiceBuilder"/> instance for method chaining </returns>
        public virtual ServiceBuilder apiSecret(string apiSecret) {
            Debug.Assert(apiSecret.Length > 0, "Invalid Api secret");
            this.apiSecret_Renamed = apiSecret;
            return this;
        }

        /// <summary>
        /// Configures the OAuth scope. This is only necessary in some APIs (like Google's).
        /// </summary>
        /// <param name="scope"> The OAuth scope </param>
        /// <returns> the <seealso cref="ServiceBuilder"/> instance for method chaining </returns>
        public virtual ServiceBuilder scope(string scope) {
            Debug.Assert(scope.Length > 0, "Invalid OAuth scope");
            this.scope_Renamed = scope;
            return this;
        }

        /// <summary>
        /// Configures the signature type, choose between header, querystring, etc. Defaults to Header
        /// </summary>
        /// <param name="type"> signature type </param>
        /// <returns> the <seealso cref="ServiceBuilder"/> instance for method chaining </returns>
        public virtual ServiceBuilder signatureType(SignatureType type) {
            this.signatureType_Renamed = type;
            return this;
        }

        public virtual ServiceBuilder debugLogger(OAuthLogger logger) {
            Debug.Assert(logger != null, "debug stream can't be null");
            this.debugLogger_Renamed = logger;
            return this;
        }

        /*
        public ServiceBuilder debug() {
            this.debugLogger(System.out);
            return this;
        }
        */

        /// <summary>
        /// Returns the fully configured <seealso cref="OAuthService"/>
        /// </summary>
        /// <returns> fully configured <seealso cref="OAuthService"/> </returns>
        public virtual OAuthService build() {
            Debug.Assert(api != null, "nullness"); // You must specify a valid api through the provider() method"
            Debug.Assert(apiKey_Renamed != null, "nullness"); // You must provide an api key
            Debug.Assert(apiSecret_Renamed != null, "nullness"); // You must provide an api secret
            return api.createService(new OAuthConfig(apiKey_Renamed, apiSecret_Renamed, callback_Renamed, signatureType_Renamed, scope_Renamed, debugLogger_Renamed));
        }
    }

}